File: ds_and_algo/implementations/javascript/data-structures/array/array.js
/*
# Built-in Array Methods in JavaScript 

- array[index] = item                           # Assigns/Updates a item of a particular element at index. If the index is out of bounds, it expands the array without error.
- array[index]                                  # Returns the item at a given index. Returns undefined if the index is out of bounds.
- array.indexOf(item)                           # Returns the index of the first occurrence of a item, or -1 if not found. 
- array.unshift(item)                           # Inserts a new element at the beginning of the array and returns the new length.
- array.splice(start, deleteCount, item1, ...)  # Add or remove elements from a start position onwards. The deleteCount defines the number of elements to remove. Returns removed elements or [].
- array.push(item1, item2..)                    # Push/Append one or more elements to the end of the array and returns the new length.
- array.shift()                                 # Removes the first element and returns the removed element. 
- array.pop()                                   # Removes the last element and returns the removed element, or undefined if the array is empty.
- array.length                                  # Returns the number of elements in the array.
- array.concat(array2, array3, …)               # Returns a new array by merging multiple arrays.
- array.slice(start, end)                       # Returns a shallow copy of a portion of the array (excluding end).
- array.reverse()                               # Reverses the array in place and returns the reversed array.
- array.sort(compareFn)                         # Sorts the array in place using the provided comparator function.
- array.map(callbackFn)                         # Returns a new array with the results of applying a function to each element.
- array.forEach(callbackFn)                     # Iterates over each element, executing a function. Returns undefined.
*/

/**
 * Class to build a dynamic Array data structure using an object.
 * @template AnyType - Generic type parameter for the elements stored in the Array (e.g., number, string, object).
 */
class CustomArray {
  #container; // Private field to hide complex implementation details (abstraction).

  /**
   *
   * @constructs CustomArray
   * @property {{[key: number]: AnyType}} container - Internal storage (object-based) to hold the elements of the array,
   * where keys are array indices. Avoids native array behavior, which defeats the purpose of learning.
   * @property {number} length - The number of elements in the array.
   */
  constructor() {
    this.#container = {}; // Faster than Object.create(null) in Chrome's V8 engine.
    this.length = 0;
  }

  /**
   *
   * @description - Access the item at the given index.
   * This method is designed to match array[index].
   * @param {number} index - The index of the element to access.
   * @returns {AnyType | undefined} - The item stored at the index, or undefined if the index is not an integer or is out of bounds.
   * @example - array.access(2)
   * Time Complexity: O(1) due to direct access by index.
   */
  access(index) {
    // Input Validation (type checking):
    if (!Number.isInteger(index)) {
      return undefined;
    }

    // Edge Case (valid negative index):
    if (index < 0) index += this.length;

    // Edge Case (if index is out of bounds):
    if (index < 0 || index >= this.length) return undefined;

    // Return the item at index:
    return this.#container[index];
  }

  /**
   *
   * @description - Search for the index of the provided item.
   * This method is designed to match array.indexOf(item).
   * @param {AnyType} item - The item to search for.
   * @returns {number} - The index of the matching item, or -1 if not found.
   * @example - array.search("item1")
   * Time Complexity: O(n) since it requires traversing to the index that matches the given item.
   */
  search(item) {
    // Loop over all elements in the Array:
    for (let i = 0; i < this.length; i++) {
      if (this.#container[i] === item) return i; // Check for matching items and return the index.
    }

    return -1;
  }

  /**
   *
   * @description - Inserts an element at a specified position, shifting existing elements to the right.
   * This method is designed to match array.splice(index, 0, item).
   * @param {number} index - The position where the new item should be inserted.
   * @param {AnyType} item - The element to be inserted at the specified index.
   * @returns {number} - The new length of the array.
   * @example - array.insert(0, "item0")
   * Time Complexity: O(n) in the worst case, as shifting elements requires linear time.
   */
  insert(index, item) {
    // Type coercion:
    index = Math.trunc(index);

    // Input validation:
    if (Number.isNaN(index)) index = 0;

    // Edge Case (if index is negative):
    if (index < 0) {
      index = this.length + index; // Normalize the negative index relative to the end.
      if (index < 0) index = 0; // If still negative, insert at the start.
    }

    // Edge Case (if index is out of bounds):
    if (index >= this.length) {
      return this.push(item); // If index exceeds the current length, insert at the end.
    }

    // From the index position onwards, shift all elements to the right (update container indices):
    this.#rightShift(index);
    // Insert the item at index:
    this.#container[index] = item;
    // Increment the length:
    this.length++;

    return this.length;
  }

  /**
   *
   * @description - Deletes an element at a specific index, shifting existing elements to the left.
   * This method is akin to array.splice(index, 1), but returns undefined instead of [].
   * @param {number} index - The index of the element to delete.
   * @returns {AnyType | undefined} - The removed element, or undefined if array is empty or index is out of bounds.
   * @example - array.delete(2)
   * Time Complexity: O(n) in the worst case, as shifting elements requires linear time.
   */
  delete(index) {
    // Input Validation (type checking):
    if (!Number.isInteger(index)) {
      return undefined;
    }

    // Edge Case (if the array is empty):
    if (this.length === 0) return undefined;

    // Edge Case (valid negative index):
    if (index < 0) {
      index = this.length + index; // Normalize negative indexes relative to the end.
      if (index < 0) index = 0; // If still negative, delete the first element.
    }

    // Edge Case (if index is out of bounds):
    if (index >= this.length) return undefined;

    // Save the element to be removed in memory:
    const removedElement = this.#container[index];

    // Edge Case (if index is the last one):
    if (index === this.length - 1) {
      delete this.#container[this.length - 1]; // Remove the last element. O(1) time complexity.
      this.length--;
      return removedElement;
    }

    // From the index position onwards, shift all elements to the left:
    this.#leftShift(index); // O(n) time complexity.

    // Decrement the length:
    this.length--;

    // Returns the removed element:
    return removedElement;
  }

  /**
   *
   * @description - Removes a specific item from the array.
   * @param {AnyType} item - The item to remove.
   * @returns {AnyType | undefined} - The removed item, or undefined if not found.
   * @example - array.remove("val1")
   * Time Complexity: O(n) in the worst case, as shifting elements requires linear time.
   */
  remove(item) {
    // Get the index of the item to be removed:
    const index = this.search(item);

    // Edge Case (if item is not found):
    if (index === -1) return undefined;

    // Remove the item at the index and return the removed item:
    return this.delete(index); // O(n).
  }

  /**
   *
   * @description - Append (push) one or more items to the end of the Array.
   * This method is designed to match array.push(item) with support to multiple arguments.
   * @param {...AnyType} items - The items to be added to the end of the Array.
   * @returns {number} - The new length of the array.
   * @example - array.push("item1", "item2")
   * Time Complexity: O(1).
   */
  push(...items) {
    // Loop over all arguments:
    for (const item of items) {
      // Insert the item at the end:
      this.#container[this.length] = item;

      // Increment the length:
      this.length++;
    }

    // Return the new length of the Array:
    return this.length;
  }

  /**
   *
   * @description - Removes the last element from the array.
   * This method is designed to match array.pop().
   * @returns {AnyType | undefined} - The removed element, or undefined if the array is empty.
   * @example - array.pop()
   * Time Complexity: O(1) for removing the last element.
   */
  pop() {
    // Edge Case (if the array is empty):
    if (this.length === 0) return undefined;

    // Save the last element in memory:
    const removedElement = this.#container[this.length - 1];

    // Remove the last element:
    delete this.#container[this.length - 1];

    // Decrement the length:
    this.length--;

    // Return the removed element:
    return removedElement;
  }

  /**
   *
   * @description - Get the internal container of the Array. Useful for debugging.
   * @returns {{[key: number]: AnyType}} - An object containing the elements of the Array.
   */
  get container() {
    return this.#container;
  }

  /**
   *
   * @private
   * @description - Shifts all elements from the given index to the right by one position.
   * @param {number} index - The index at which to shift elements.
   * Time Complexity: O(n) in the worst case.
   */
  #rightShift(index) {
    // Loop backwards through the Hash Table from self.length until the index position:
    for (let i = this.length; i > index; i--) {
      this.#container[i] = this.#container[i - 1];
    }
  }

  /**
   *
   * @private
   * @description - Shifts all elements from the given index to the left by one position.
   * @param {number} index - The index at which to shift elements.
   * Time Complexity: O(n) in the worst case.
   */
  #leftShift(index) {
    // Loop forward through the Hash Table from the index position until the second-to-last element:
    for (let i = index; i < this.length - 1; i++) {
      this.#container[i] = this.#container[i + 1]; // Left shift elements.
    }

    // Delete the last duplicate element:
    delete this.#container[this.length - 1];
  }

  /**
   * @description - Iterable protocol implementation to allow for...of loops, spread, etc.
   * @returns {object} - An iterator object.
   */
  [Symbol.iterator]() {
    let i = 0;
    const length = this.length;
    const container = this.#container;

    // Return an iterator object with a 'next' method using object literal and arrow function syntax.
    // Each call to next() returns an object of the form { value, done }.
    return {
      next: () =>
        i < length ? { value: container[i++], done: false } : { done: true },
    };
  }
}

// Example usage:
if (import.meta.url.endsWith(process.argv[1])) {
  let arr = new CustomArray();
  console.log(arr.container); // output: {}
  for (let i = 0; i < 5; i++) {
    arr.push(i + 1);
  }
  console.log(arr.container); // output: { '0': 1, '1': 2, '2': 3, '3': 4, '4': 5 }
  arr.insert(-5, "item0");
  console.log([...arr]); // output: ["item0", 1, 2, 3, 4, 5]
  console.log(arr.access(0)); // output: "item0"
  console.log(arr.search("item0")); // output: 0
  arr.delete(0);
  console.log([...arr]); // output: [1, 2, 3, 4, 5]
  arr.remove(5);
  console.log([...arr]); // output: [1, 2, 3, 4]
  arr.pop();
  console.log([...arr]); // output: [1, 2, 3]
  console.log(arr.length); // output: 3
}

File: ds_and_algo/implementations/python/data_structures/array/array.py
"""Examples of built-in list operations in Python.

Notes:
    - array[index] = item: Assign or update an item by index.
    - array[index]: Return the item at a given index.
    - array.index(item): Return the index of the first matching item.
    - array.insert(index, item): Insert at an index and shift elements right.
    - del array[index]: Delete an item and shift elements left.
    - array.remove(item): Remove the first matching item.
    - array.append(item): Append a new item to the end of the array.
    - array.pop(index): Remove and return the item at an index.
"""

from typing import Generic, TypeVar

AnyType = TypeVar("AnyType")  # Array items can be of any data type.


class Array(Generic[AnyType]):
    """
    Class to build a Dynamic Array using a list as a data container.

    Attributes:
        _capacity: The maximum number of elements the array can hold.
        _container: Internal storage (list-based) to hold
            the elements of the array, where indices are array positions.
        _length: The number of elements in the array.
    """

    _capacity: int
    _container: list[AnyType | None]
    _length: int

    def __init__(self) -> None:
        """Initialize an empty array."""

        self._capacity = 10
        self._container = [None] * self._capacity
        self._length = 0

    def _resize(self):
        self._capacity *= 2
        new: list[AnyType | None] = [None] * self._capacity

        for i in range(self._length):
            new[i] = self._container[i]

        self._container = new

    def access(self, index: int) -> AnyType:
        """
        Access the item at the given index.

        Args:
            index: The index of the item to access.

        Returns:
            The item at the index.

        Raises:
            TypeError: If index is not an integer.
            IndexError: If index is out of range.

        Notes:
            Time complexity: O(1).
        """

        # Input Validation (type checking):
        if type(index) is not int:
            raise TypeError(f"Index must be an integer, got {type(index).__name__}.")

        # Edge Case (valid negative index):
        if index < 0:
            index += self._length

        # Edge Case (if the Array is empty or index is out of range):
        if index < 0 or index >= self._length:
            raise IndexError("Array index out of range.")

        # Return the item at index:
        return self._container[index]  # type: ignore

    def search(self, item: AnyType) -> int:
        """
        Search for the index of the provided item.

        Args:
            item: The item to search for.

        Returns:
            The index of the matching item.

        Raises:
            ValueError: If the item is not found in the array.

        Notes:
            Time complexity: O(n).
        """

        # Loop over all elements in the Array:
        for i in range(self._length):
            if self._container[i] == item:  # Check for matching items.
                return i  # Return the index.

        raise ValueError(f"{item} is not in Array.")

    def insert(self, index: int, item: AnyType) -> None:
        """
        Insert an item at an index and shift elements to the right.
        This method is akin to the built-in array.insert(index, item) operation.

        Args:
            index: The position where the item should be inserted.
            item: The item to insert.

        Notes:
            Time complexity: O(n) in the worst case.
        """

        # Input Validation (type checking):
        if type(index) is not int:
            raise TypeError(f"Index must be an integer, got {type(index).__name__}.")

        # Edge Case (if index is negative):
        if index < 0:
            index += self._length  # Normalize negative indices.
            if index < 0:
                index = 0  # If the normalized index is still negative, set it to 0.

        # Edge Case (if index is out of range):
        if index >= self._length:
            # If index exceeds the current length, insert at the end:
            self.append(item)
        else:
            if self._length == self._capacity:
                self._resize()

            # From the index position onwards, shift all elements to the right
            # (update container indices):
            self.__right_shift(index)

            # Insert the item at index:
            self._container[index] = item

            # Increment the length:
            self._length += 1

    def delete(self, index: int) -> None:
        """
        Delete the element at an index and shift elements to the left.
        This method is akin to the built-in del array[index] operation.

        Args:
            index: The index of the element to delete.

        Raises:
            IndexError: If the array is empty or the index is out of range.

        Notes:
            Time complexity: O(n) in the worst case.
        """

        # Input Validation (type checking):
        if type(index) is not int:
            raise TypeError(f"Index must be an integer, got {type(index).__name__}.")

        # Edge Case (if the Array is empty):
        if self._length == 0:
            raise IndexError("List assignment index out of range!")

        # Edge Case (valid negative index):
        if index < 0:
            index += self._length

        # Edge Case (if index is out of range):
        if index < 0 or index >= self._length:
            raise IndexError("List index out of range!")

        # Edge Case (if removing the last element):
        if index == self._length - 1:
            self._container[index] = None  # Remove the last element. O(1) time.
        else:
            # Otherwise, from the index position, shift all elements to the left:
            self.__left_shift(index)  # O(n) time complexity.

        # Decrement the length:
        self._length -= 1

    def remove(self, item: AnyType) -> None:
        """
        Remove the first matching item from the array.

        Args:
            item: The item to remove.

        Raises:
            ValueError: If the item is not found in the array.

        Notes:
            Time complexity: O(n) in the worst case.
        """

        index = self.search(item)  # This will raise ValueError if item is not found.
        self.delete(index)  # O(n).

    def append(self, item: AnyType) -> None:
        """
        Append an item to the end of the array.

        Args:
            item: The item to append.

        Notes:
            Time complexity: O(1).
        """

        # Resize if full:
        if self._length == self._capacity:
            self._resize()

        # Insert the item at the end:
        self._container[self._length] = item

        # Increment the length:
        self._length += 1

    def pop(self) -> AnyType:
        """
        Remove and return the last item.

        Returns:
            The last item in the array.

        Raises:
            IndexError: If the array is empty.

        Notes:
            Time complexity: O(1).
        """

        # Edge Case (if the Array is empty):
        if self._length == 0:
            raise IndexError("Cannot pop from empty Array!")

        # Save the last item in memory:
        removed_item = self._container[self._length - 1]

        # Remove the last item from the array:
        self._container[self._length - 1] = None

        # Decrement the length:
        self._length -= 1

        # Return the removed item:
        return removed_item  # type: ignore

    def __delitem__(self, index: int) -> None:
        """Enables the `del array[index]` syntax to delete an element at an index."""
        self.delete(index)

    def __right_shift(self, index: int) -> None:
        """
        Shift elements to the right starting at the given index.

        Args:
            index: The position from which to shift elements.

        Notes:
            Time complexity: O(n).
        """

        # Loop backwards through the Hash Table until the index position:
        for i in range(self._length, index, -1):
            self._container[i] = self._container[i - 1]  # Right shift elements.

    def __left_shift(self, index: int) -> None:
        """
        Shift elements to the left starting at the given index.

        Args:
            index: The position from which to shift elements.

        Notes:
            Time complexity: O(n).
        """

        # Loop forward through the Hash Table from the index position until the
        # second-to-last element:
        for i in range(index, self._length - 1):
            self._container[i] = self._container[i + 1]  # Left shift elements.

        # Remove the last duplicated item:
        self._container[self._length - 1] = None

    def __len__(self) -> int:
        """Return the number of elements and enables `len(array)` syntax."""
        return self._length

    def __repr__(self) -> str:
        """Return a string representation of the array."""
        return str([self._container[i] for i in range(self._length)])


# Example usage:
if __name__ == "__main__":
    array = Array()
    print(array)  # []
    for i in range(5):
        array.append(i + 1)
    print(array)  # [1, 2, 3, 4, 5]
    array.insert(-5, "item0")
    print(array)  # ['item0', 1, 2, 3, 4, 5]
    print(array.access(0))  # item0
    print(array.search("item0"))  # 0
    array.delete(0)
    print(array)  # [1, 2, 3, 4, 5]
    array.remove(5)
    print(array)  # [1, 2, 3, 4]
    array.pop()
    print(array)  # [1, 2, 3]
    print(array._capacity)  # 10
    print(len(array))  # 3
